### PR TITLE
Accumulated gyro filter fixes and improvements

### PR DESCRIPTION
**Pull request by dagar**
See commits for details.

##### Baseline (fake IMU sine sweep ramping to 100 Hz over 10 seconds)
Screenshot from 2021-05-25 13-57-09

#####  ESC dynamic notch filtering enabled (with intentional gaps in data at different points for testing resets and timeouts)
Screenshot from 2021-05-25 13-59-45

##### FFT dynamic notch filtering enabled (still too much latency in the FFT peak detection, but more to come)
Screenshot from 2021-05-25 13-55-06

### CODE REVIEW

**Comment by dagar** [comment]
I'm working on moving the `gyro_fft` to a sliding DFT and away from CMSIS entirely, but it's not quite ready yet. This will improve the output rate considerably.

**Review by bresch — CHANGES_REQUESTED** [review]

**Comment by dagar** [comment]
New idea with filters
 - be paranoid checking that everything is finite during filter parameter change or reset
 - move finite checks out of filter updates entirely (they're not free and this is a hot path)
 - push finite checks out to users where we're not sure
      -  for example we're not going to get NANs in the expensive sensor_gyro_fifo case

Thoughts?

**Comment by dagar** [comment]
I've updated the NotchFilter such that changes to only the notch frequency will be gradually applied, shifting it no more than half the bandwidth per filter update.

**Comment by bresch** [comment]
> New idea with filters
> - be paranoid checking that everything is finite during filter parameter change or reset
> - move finite checks out of filter updates entirely (they're not free and this is a hot path)
>  - push finite checks out to users where we're not sure
        - for example we're not going to get NANs in the expensive sensor_gyro_fifo case

Yes, I'm fine with that

File: src/lib/mathlib/math/filter/NotchFilter.hpp [hunk 1]
```diff
@@ -109,8 +86,17 @@ class NotchFilter
 		return output;
 	}
 
-	float getNotchFreq() const { return _notch_freq; }
+	// Filter array of samples in place using the direct form I
+	inline void applyArray(T samples[], int num_samples)
+	{
+		for (int n = 0; n < num_samples; n++) {
+			samples[n] = apply(samples[n]);
```
[the remark this answers is not in the document]
Comment (dagar) [reply src/lib/mathlib/math/filter/NotchFilter.hpp:93]: After some quick profiling I'd say it's a bit questionable if there's potential for larger jumps in notch frequency. Updating parameters for only one NotchFilter in an iteration is quite expensive, and there are quite a few of them potentially active.

Updated to instead reset each filter if the notch frequency shifts by more than the bandwidth.

**Review by bresch — APPROVED** [review]

**Pull request merged**